File: Algorithms/0114_Flatten_Binary_Tree_to_Linked_List/Java/FlattenBinaryTreetoLinkedListSolution2.java

// Space: O(1)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public void flatten(TreeNode root) {
        if (root == null) return;
        if (root.left == null & root.right == null) return;
        dfs(root);
    }

    private TreeNode dfs(TreeNode root) {
        if (root == null) return null;
        TreeNode leftLast = dfs(root.left);
        TreeNode rightLast = dfs(root.right);
        if (leftLast != null) {
            leftLast.right = root.right;
            root.right = root.left;
            root.left = null;
        }
        if (rightLast != null) return rightLast;
        if (leftLast != null) return leftLast;
        return root;
    }
}

File: Algorithms/0114_Flatten_Binary_Tree_to_Linked_List/Python/Flatten_Binary_Tree_to_Linked_List_Solution_2.py

# Space: O(1)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def flatten(self, root):
        """
        Do not return anything, modify root in-place instead.
        """
        if root is None: return
        if root.left is None and root.right is None: return

        def dfs(root):
            if root is None: return None
            leftlast = dfs(root.left)
            rightlast = dfs(root.right)

            if leftlast is not None:
                leftlast.right = root.right
                root.right = root.left
                root.left = None

            if rightlast: return rightlast
            if leftlast: return leftlast
            return root

        dfs(root)




